### ISSUE TITLE
Severe JobRunr Exception

### ISSUE THREAD

**Issue by zhemaituk** [jobrunr/jobrunr#186]
Reproduced with JobRunr v3.2.1.

### Runtime information
- __Timestamp__: 2021-07-30T19:49:07.000088Z
- __Location__: BackgroundJobServer 88427d46-caf0-4fc7-b618-231d8f239228
- __JobRunr Version__: Unable to determine version
- __StorageProvider__: org.jobrunr.storage.sql.postgres.PostgresStorageProvider
- __Java Version__: 11.0.12
- __Is running from nested jar__: true

### Background Job Servers
- BackgroundJobServer id: 88427d46-caf0-4fc7-b618-231d8f239228
(workerPoolSize: 32, pollIntervalInSeconds: 5, firstHeartbeat: 2021-07-30T14:09:50.877704Z, lastHeartbeat: 2021-07-30T19:49:05.878348Z)

#### Concurrent modified jobs:
Job id: bfd8886e-e31b-484f-b131-190c1610617a
	Local version: 3; Storage version: 1
	Local state: DELETED (at 2021-07-30T19:49:06.975142Z) ← SUCCEEDED (at 2021-07-30T19:49:03.242117Z) ← PROCESSING (at 2021-07-30T19:49:02.447094Z)
	Storage state: ENQUEUED (at 2021-07-30T19:49:06.977461Z)

### Exception
```java
org.jobrunr.SevereJobRunrException: Could not resolve ConcurrentJobModificationException
	at org.jobrunr.server.JobZooKeeper.processJobList(JobZooKeeper.java:207)
	at org.jobrunr.server.JobZooKeeper.processJobList(JobZooKeeper.java:191)
	at org.jobrunr.server.JobZooKeeper.checkForSucceededJobsThanCanGoToDeletedState(JobZooKeeper.java:129)
	at org.jobrunr.server.JobZooKeeper.runMasterTasksIfCurrentServerIsMaster(JobZooKeeper.java:95)
	at org.jobrunr.server.JobZooKeeper.run(JobZooKeeper.java:75)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

```

**Comment by rdehuyss** [comment]
Please upgrade to the latest version, this is already solved.

**Comment by rdehuyss** [comment]
P.s.: I saw you already created an issue for this before; see https://github.com/jobrunr/jobrunr/issues/138.

Any reason why you did not upgrade back then?

**Comment by rdehuyss** [comment]
I'm sorry I closed the issue so fast - normally with JobRunr 3.2.1, it would give the correct version where it now says unable to determine version.

I'll look into it.

**Comment by rdehuyss** [comment]
Question: do you delete jobs situationally after they succeed or was the job deleted manually just after it finished?

**Comment by zhemaituk** [comment]
@rdehuyss thanks for the attention to this one. all good questions.

- does this happen often?
Yes. Under load of about a thousand jobs per minute it happens a few times a day. Usually less than five, so it went unnoticed for some time, till one day it actually halted.

- what is the setting of deleteSucceededJobsAfter? I assume it is really low?
it's 0. full config:
```
org:
  jobrunr:
    job-scheduler:
      enabled: true
    background-job-server:
      enabled: true
      poll-interval-in-seconds: 5
      worker_count: 32
      delete-succeeded-jobs-after: 0s
      permanently-delete-deleted-jobs-after: 0s
    dashboard:
      enabled: true
```

- do you use StorageProvider.deletePermanently in your code?
Yes.
There is a hack-y approach to make jobs not to build up, if "identical" job didn't finish yet.
There must be a better way to achieve it, but maybe you will consider this approach as not totally insane. I considered mutexes for this, but it also wouldn't prevent the jobs queueing up.

Such code is executed every minute (uuidList is more or less static during the day, and is about 1000 records in size):
```
for (UUID jobId : uuidList) {
            Job job;
            try {
                job = storageProvider.getJobById(tradingId);
            } catch (JobNotFoundException e) {
                job = null;
            }

            if (job != null) {
                if (ACTIVE_JOB_STATES.contains(job.getState())) {
                    log.warn("Skipping job {} for current minute since previous minute launch didn't complete yet.", tradingId);
                    continue;
                } else {
                    storageProvider.deletePermanently(jobId);
                }
            }

            jobScheduler.enqueue(
                    jobId,
                    () -> processorLauncher.process(jobId)
            );
        }
```

- are you manually specifying the id's when creating jobs?
Yes

- Are you using a time machine/traveling faster than light? 😄 The reason I don't understand is that the newly enqueued job has a timestamp newer than the succeeded state. So for JobRunr it appears that it succeeded before it was created.
I understand the approach taken is on the edge of the how the JobRunr is intended to be used, but it was doing the job until load went higher.

**Comment by rdehuyss** [comment]
Can you perhaps reach out via LinkedIn as I'm looking for a way to quickly chat about this, to get some more info about the business context.

https://www.linkedin.com/in/ronalddehuysser/

Thx!

**Comment by rdehuyss** [comment]
@zhemaituk I realised this is solved in the Pro version using the [Custom Delete Policies](https://www.jobrunr.io/en/documentation/pro/custom-delete-policy/) where the the Job is instantly deleted.

**Comment by rdehuyss** [comment]
Hi @zhemaituk , I just released JobRunr 4.0.0-RC1 with some fixes for this. Can you test and report back?

**Comment by zhemaituk** [comment]
Have been using 4.0.0 release for a week - all good, the issue is not reproducible. Thank you Ronald.
